blocks/api/raw-handling/list-merger.js

```diff
@@ -14,16 +20,40 @@ export default function( node ) {
 		return;
 	}
 
-	if ( node.children.length !== 1 ) {
-		return;
-	}
-
+	const list = node;
 	const prevElement = node.previousElementSibling;
 
-	if ( ! prevElement || prevElement.nodeName !== type ) {
-		return;
+	// Merge with previous list if:
+	// * There is a previous list of the same type.
+	// * There is only one list item.
+	if (
+		prevElement &&
+		prevElement.nodeName === type &&
+		list.children.length === 1
+	) {
+		prevElement.appendChild( list.firstChild );
+		list.parentNode.removeChild( list );
 	}
 
-	prevElement.appendChild( node.firstChild );
-	node.parentNode.removeChild( node );
+	const parentElement = node.parentNode;
+
+	// Nested list with empty parent item.
+	if (
+		parentElement &&
+		parentElement.nodeName === 'LI' &&
+		parentElement.children.length === 1 &&
+		! /\S/.test( shallowTextContent( parentElement ) )
+	) {
+		const parentListItem = parentElement;
+		const prevListItem = parentListItem.previousElementSibling;
+		const parentList = parentListItem.parentNode;
+
+		if ( prevListItem ) {
+			prevListItem.appendChild( list );
+			parentList.removeChild( parentListItem );
+		} else {
+			parentList.parentNode.insertBefore( list, parentList );
+			parentList.parentNode.removeChild( parentList );
+		}
+	}
```

Comment:
Is the name `listMerger` still adequate? `listReducer`? `Condenser`? These are intentionally more abstract (less loaded). Alternatively, does it make sense at some point to split and have two filters, `listMerger` and `listFlattener`?

Reply:
Yeah these names start to make less sense for a let of modules... :) We also have "corrector" and "normaliser".

Reply:
I'm not great at naming such things. I don't mind renaming almost everything to "reducer"... :)